File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.java
import java.util.*;
import java.io.*;

class Solution {
    public int[] solution(String[] operations) {
        int[] answer = {};
        //오름차순 우선순위 큐 생성
        PriorityQueue<Integer> minpq = new PriorityQueue<>();
        // 내림차순 우선순위 큐 생성
        PriorityQueue<Integer> maxpq = new PriorityQueue<>(Collections.reverseOrder());
        
        for(String str : operations){
            if(str.split(" ")[0].equals("I")){
                minpq.offer(Integer.parseInt(str.split(" ")[1]));
                maxpq.offer(Integer.parseInt(str.split(" ")[1]));
            } 
            else if(str.split(" ")[0].equals("D")){
                //최댓값 삭제
                if(str.split(" ")[1].equals("1")){
                     minpq.remove(maxpq.poll());
                }
                //최솟값 삭제
                else if(str.split(" ")[1].equals("-1")){
                    maxpq.remove(minpq.poll());
                }
            }
            
        }
        
        if(maxpq.isEmpty() && minpq.isEmpty()){
            return new int[]{0,0};
        }
        return new int[]{maxpq.poll(), minpq.poll()};
    }
}

File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.py
import heapq
from collections import Counter
def clean_max_heap(max_heap, count):
    while max_heap and count[-max_heap[0]] == 0:
        heapq.heappop(max_heap)
 
def clean_min_heap(min_heap, count):
    while min_heap and count[min_heap[0]] == 0:
        heapq.heappop(min_heap)
    
def solution(operations):
    answer = []
    min_heap = []
    max_heap = []
    count = Counter()
    
    for op in operations:
        order, num = op.split(" ")
        num = int(num)
        if(order == 'I'):
            # 큐에 주어진 숫자 삽입
            heapq.heappush(min_heap,num)
            heapq.heappush(max_heap,-num)
            count[num] += 1
        elif order == 'D':
            if num == 1:
                # 큐에서 최댓값 삭제
                clean_max_heap(max_heap, count)
                if max_heap:
                    x = -heapq.heappop(max_heap)
                    count[x] -= 1
            elif num == -1:
                # 큐에서 최솟값 삭제
                clean_min_heap(min_heap, count)
                
                if min_heap:
                    x = heapq.heappop(min_heap)
                    count[x] -= 1
                    
    # 마지막 결과 확인 전에도 무효값 정리
    clean_min_heap(min_heap, count)
    clean_max_heap(max_heap, count)

    if not min_heap and not max_heap:
        return [0,0]
    
    return [-max_heap[0], min_heap[0]]